pkg/resource/restconfig_factory.go

```diff
@@ -63,3 +181,25 @@ func (r *restConfigClientGetter) ToRawKubeConfigLoader() clientcmd.ClientConfig
 	}
 	return clientcmd.NewNonInteractiveClientConfig(*cfg, "", overrides, nil)
 }
+
+func (r *restConfigClientGetter) StillValid() bool {
+	err := validateRestConfig(r.restConfig)
+	if err != nil {
+		r.logger.WithError(err).Debug("cached rest config was invalid")
+		return false
+	}
+
+	r.logger.Debug("cached rest config was still valid")
+	return true
+}
+
+// validateRestConfig returns nil iff the rest config is valid and working.
+func validateRestConfig(config *rest.Config) error {
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	// perform a basic action (listing namespaces) to verify connection
+	_, err = clientset.CoreV1().Namespaces().List(context.Background(), metav1.ListOptions{})
```

Comment:
ee, this could be pretty expensive. If we are gonna keep a test query, I think we want it to be something cheaper, like maybe grabbing the `infrastructure` object (non-namespaced singleton named `cluster`).

But I don't love the idea of doing a test query on the spoke every time we pull a cache entry. We talked about letting the normal reconcile fail organically, detecting that that failure had to do with the cache entry being stale, and using that to invalidate it. Did you explore that option? One possible impl approach that comes to mind would involve wrapping the clustersync Reconcile func to do that discovery/invalidation before bubbling up the error.

Also, we talked about only invalidating the cache entry for _certain_ errors. You were going to figure out what it looks like when certs expire... which I _think_ is the only way this happens? Because e.g. for some temporary network glitch, as currently written we would end up thrashing, creating fresh getters, resulting in the original leak, right?